src/main/java/org/opensearch/flowframework/util/ParseUtils.java

```diff
@@ -229,6 +247,218 @@ public static User getUserContext(Client client) {
+     * @param requestedUser the user to execute the request
+     * @param workflowId workflow id
+     * @param filterByEnabled filter by enabled setting
+     * @param listener action listener
+     * @param function workflow function
+     * @param client node client
+     * @param clusterService cluster service
+     * @param xContentRegistry contentRegister to parse get response
+     */
+    public static void resolveUserAndExecute(
+        User requestedUser,
+        String workflowId,
+        boolean filterByEnabled,
+        ActionListener listener,
+        WorkflowFunction function,
+        Client client,
+        ClusterService clusterService,
+        NamedXContentRegistry xContentRegistry
+    ) {
+        try {
+            if (requestedUser == null || !filterByEnabled) {
+                // requestedUser == null means security is disabled or user is superadmin. In this case we don't need to
+                // check if request user have access to the workflow or not.
+                // !filterByEnabled means security is enabled and filterByEnabled is disabled
+                function.execute();
+            } else {
+                getWorkflow(requestedUser, workflowId, filterByEnabled, listener, function, client, clusterService, xContentRegistry);
+            }
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
+    }
+
+    /**
+     * Check if requested user has backend role required to access the resource
+     * @param requestedUser the user to execute the request
+     * @param resourceUser user of the resource
+     * @param workflowId workflow id
+     * @return boolean if the requested user has backend role required to access the resource
+     * @throws Exception exception
+     */
+    private static boolean checkUserPermissions(User requestedUser, User resourceUser, String workflowId) throws Exception {
+        if (resourceUser.getBackendRoles() == null || requestedUser.getBackendRoles() == null) {
+            return false;
+        }
+        // Check if requested user has backend role required to access the resource
+        for (String backendRole : requestedUser.getBackendRoles()) {
+            if (resourceUser.getBackendRoles().contains(backendRole)) {
+                logger.debug(
+                    "User: "
+                        + requestedUser.getName()
+                        + " has backend role: "
+                        + backendRole
+                        + " permissions to access config: "
+                        + workflowId
+                );
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Check if filter by backend roles is enabled and validate the requested user
+     * @param requestedUser the user to execute the request
+     * @return error message if validation fails, null otherwise
+     */
+    public static String checkFilterByBackendRoles(User requestedUser) {
```

Comment:
Returning a `String` or `null` and then testing that null to throw an exception seems an odd pattern.

Why not just have a void method here that throws an exception?  Most validate-style methods do this.  That would allow you to customize the RestStatus as well (403 vs. 404, etc.)